pkg/common-controller/snapshot_controller.go

```diff
@@ -171,53 +158,68 @@ func (ctrl *csiSnapshotCommonController) syncContent(content *crdv1.VolumeSnapsh
 func (ctrl *csiSnapshotCommonController) syncSnapshot(snapshot *crdv1.VolumeSnapshot) error {
 	klog.V(5).Infof("synchronizing VolumeSnapshot[%s]: %s", utils.SnapshotKey(snapshot), utils.GetSnapshotStatusForLogging(snapshot))
 
-	err := ctrl.processFinalizersAndCheckandDeleteContent(snapshot)
-	if err != nil {
-		return err
-	}
-
-	if !utils.IsSnapshotReady(snapshot) {
-		return ctrl.syncUnreadySnapshot(snapshot)
+	if snapshot.ObjectMeta.DeletionTimestamp != nil {
+		err := ctrl.processIfDeletionTimestampSet(snapshot)
+		if err != nil {
+			return err
+		}
+	} else {
+		klog.V(5).Infof("syncSnapshot[%s]: check if we should add finalizers on snapshot", utils.SnapshotKey(snapshot))
+		ctrl.checkandAddSnapshotFinalizers(snapshot)
```

Comment:
each time snapshot sync, it will trigger checkandAddSnapshotFinalizers, which will check and update finalizer even if finalizer is already added.